### PR TITLE
7472 outbound shipment edit refactor

### LINKED ISSUE #7472: Prescription line edit: simplify logic
## The suggestion

Hi triage, just wanting to make a placeholder for this work, I think its super important. Ideally should be done in amongst move back from the line edit view....

- Managing state for all items, rather than just the item in view, adds way more complexity
- There's no validation on the save button (just some very hacky places where we set know we've done something invalid, so we set "form not touched, nothing to save"/`isDirty = false`)
- Reusing allocation logic from outbound is adding complexity, use cases that are hard to be aware of
  - Probably if allocation logic was broken down into pieces, we could share elements, and leave specific logic in each area
- Concept of placeholder line/unallocated stock is different than for other invoices
  - You can have picked prescription with placeholder line, for prescribed quantity (though maybe not really anymore? honestly can't tell)
  - When should a prescription actually be in New vs Picked?
- Batch mutations are hard to reason about (one example, `isCreated` seems to be creating some strange behaviour [per this PR comment](https://github.com/msupply-foundation/open-msupply/pull/7495#pullrequestreview-2797772473))
- We need comprehensive tests for this stuff, its way too brittle and edge casey
- There are item-level things stored on prescription, which on backend we actually store against a line (there are potentially many lines to an item)
  - This adds a lot of complexity to frontend logic, having to know about this
  - I think we should go for API that supports this - e.g. can just query for item directions for this item on a prescription, rather than -> query all lines for item -> loop through -> find one with a directions (and on edit, do you update one or all lines etc...)

I did some RnD a while ago but didn't finish: https://github.com/msupply-foundation/open-msupply/pull/6817

### PR DESCRIPTION
**Pull request by lache-melvin**
Part of #7472

IS BEAST HEH - plz just grab me im happy to chat through changes 🙈 ❤️ 

# 👩🏻‍💻 What does this PR do?

Only UI change is the removal of pack selector here - discussion with PMs was that it wasn't used, and adds confusion. Given there was a lot of re-implementation here, was easier to leave off for now. If it turns out it is a required feature, we can add back separately:

Screenshot 2025-05-13 at 12 43 36 PM

OTHERWISE - epic rearrangement + simplification of code.

Included:
- Targeted query & mutation for the outbound edit modal (allows removal of batch (isCreated/updated logic)
- Introduces `allocationContext` - all allocation logic and state handled through here, one place
  - This will be where you pass your "lens" (issue in doses or packs, sort by expiry of vvm etc.)
- Handle placeholder entirely separate, rather than always looking for placeholder line in arrays
- Bit of a rearrange of components based on removal of dependencies

## 💌 Any notes for the reviewer?

Sorry for how large - I think for big refactor like this, hard to do piece by piece without quite a lot of churn... anyway, would probably recommend reviewing by just navigating the code that's there now, than trying to parse the changes?

Haven't refactored prescriptions - would like to do this over cooldown period, to reuse this new AllocationContext. Would be able to delete a lot more code after that, there are a lot of shared dependencies.

Also still to do, maybe in cool down? lots to try get in for 2.8, feel I shouldn't spend too much more time here this week: document expected behaviour - for discussion with product team

# 🧪 Testing

- [ ] Barring the removal of the pack selector, it should work the same. Please test:
- [ ] Auto allocation:
  - [ ] Allocates correct quantity
  - [ ] Splits across many stock lines
  - [ ] Warns if not enough stock (creates placeholder in New status, otherwise just warn)
  - [ ] Rounds up to nearest pack size (e.g. if requesting 50 but you only have packs of 100)
  - [ ] Doesn't allocate expired/on hold
  - [ ] Allocates nearest expiring first
- [ ] Can manually allocate expired lines, but not on hold
- [ ] Manual allocation updates the auto-allocation value
- [ ] Add from scanner button on android (scanned batch is highlighted, others are disabled)

# 📃 Documentation

- [ ] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [x] **These areas should be updated or checked**: 
  1. Screenshots & mentions of packs
  2.

# 📃 Reviewer Checklist

The PR Reviewer(s) should fill out this section before approving the PR

**Breaking Changes**
- [ ] No Breaking Changes in the Graphql API
- [ ] Technically some Breaking Changes but not expected to impact any integrations

**Issue Review**
- [ ] All requirements in original issue have been covered
- [ ] A follow up issue(s) have been created to cover additional requirements

**Tests Pass**
- [ ] Postgres
- [ ] SQLite
- [ ] Frontend

### CODE REVIEW

File: client/packages/common/src/utils/arrays/SortUtils.ts [hunk 1]
```diff
@@ -26,8 +26,13 @@ const parseValue = <T, K extends keyof T>(object: T, key: K) => {
 };
 
 const getSortByExpiry =
-  (desc = false) =>
+  (desc = false, nonExpiringLast = false) =>
```
Comment (lache-melvin) [inline client/packages/common/src/utils/arrays/SortUtils.ts:29]: Old way was leaving the lines with no expiry value in place? Probably mostly a dev data problem, but I would think "non-expiring" batches should be allocated last

File: client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx [hunk 2]
```diff
@@ -13,12 +13,12 @@ import {
...
 import { useOutbound } from '../api';
 import { isOutboundDisabled } from '../../utils';
-import { Draft } from '../..';
+import { ScannedBarcode } from '../../types';
 
 export const AddFromScannerButtonComponent = ({
   onAddItem,
 }: {
-  onAddItem: (draft?: Draft) => void;
+  onAddItem: (scannedBarcode?: ScannedBarcode) => void;
```
Comment (lache-melvin) [inline client/packages/invoices/src/OutboundShipment/DetailView/AddFromScannerButton.tsx:21]: updated the onOpen logic to explicitly handle opened via barcode vs open via item click

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/Allocation.tsx [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+} from './allocation/useAllocationContext';
+import { sumAvailableQuantity } from './allocation/utils';
+
+interface AllocationProps {
+  itemData: OutboundLineEditData;
+  allowPlaceholder: boolean;
+  scannedBatch?: string;
+}
+
+export const Allocation = ({
```
Comment (lache-melvin) [inline client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/Allocation.tsx:32]: Please have a gander per new component structure - but idea was:

```
- OutboundLineEdit (the modal)
  - SelectItem
  - Allocation
    - AutoAllocation (input + alerts)
    - Table (manual allocation)
```

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/Allocation.tsx [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+}: AllocationProps) => {
+  const { initialise, initialisedForItemId } = useAllocationContext(
+    ({ initialise, initialisedForItemId }) => ({
+      initialise,
+      initialisedForItemId,
+    })
+  );
+
+  useEffect(() => {
+    initialise(
```
Comment (lache-melvin) [inline client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/Allocation.tsx:45]: Wrapper component... one time initialisation of the allocation context per item

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx [hunk 5]
```diff
@@ -1,197 +1,130 @@
...
 
-  const placeholder = draftStockOutLines?.find(
-    ({ type, numberOfPacks }) =>
-      type === InvoiceLineNodeType.UnallocatedStock && numberOfPacks !== 0
+  const { data: itemData, isFetching } = useOutboundLineEditData(
+    invoiceId,
+    itemId
   );
+  const { mutateAsync } = useSaveOutboundLines(invoiceId);
+  const { saveBarcode } = useOpenedWithBarcode(asBarcodeOrNull(openedWith));
```
Comment (lache-melvin) [inline client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx:58]: Pulled as much of the barcode-y biz into this hook as i could

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx [hunk 6]
```diff
@@ -1,197 +1,130 @@
...
   const handleSave = async (onSaved: () => boolean | void) => {
+    const confirmZeroQuantityMessage = t('messages.confirm-zero-quantity');
     if (
-      getAllocatedQuantity(draftStockOutLines) === 0 &&
-      !showZeroQuantityConfirmation
+      allocatedUnits === 0 &&
+      !alerts.some(alert => alert.message === confirmZeroQuantityMessage)
     ) {
-      setShowZeroQuantityConfirmation(true);
+      setAlerts([{ message: confirmZeroQuantityMessage, severity: 'warning' }]);
```
Comment (lache-melvin) [inline client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx:103]: Didn't want to have to pass the showZeroQuantityConfirmation prop all the way through, and I think this is more semantic - just set the alerts in the context to whatever they need to be, and let the renderer display them.

I think a further improvement would be for the translation of the alert messages to be in the component, so we're just dealing with the logical concept of the alert at this point but... now I'm in the weeds 😁

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx [hunk 7]
```diff
@@ -1,197 +1,130 @@
...
-  const { warning } = useNotification();
-  const [showZeroQuantityConfirmation, setShowZeroQuantityConfirmation] =
-    useState(false);
-  useFocusNumberOfPacksInput(draft);
+  const { next, disabled: nextDisabled } = useNextItem(itemId);
 
-  const placeholder = draftStockOutLines?.find(
-    ({ type, numberOfPacks }) =>
-      type === InvoiceLineNodeType.UnallocatedStock && numberOfPacks !== 0
+  const { data: itemData, isFetching } = useOutboundLineEditData(
```
Comment (lache-melvin) [inline client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx:53]: custom query - all the bits we need for the modal in one query - rather than composing several queries, with loads of state updates after each resolves. Much more efficient for slower networks/devices too 🎉

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/allocateQuantities.test.ts [hunk 8]
```diff
@@ -0,0 +1,186 @@
+import { FnUtils } from '@openmsupply-client/common';
+import { allocateQuantities } from './allocateQuantities';
+import { DraftStockOutLineFragment } from '../../../api/operations.generated';
+
+describe('allocateQuantities - standard behaviour.', () => {
```
Comment (lache-melvin) [inline client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/allocateQuantities.test.ts:5]: test cases are the same as old tests (removed no-longer-relevant cases like placeholder though)

File: client/packages/invoices/src/OutboundShipment/api/hooks/utils/index.ts [hunk 9]
```diff
@@ -5,7 +5,6 @@ import { useAddFromMasterList } from './useAddFromMasterList';
 import { useBarcode } from './useBarcode';
 import { useBarcodeInsert } from './useBarcodeInsert';
 import { useSelectedLines } from './useSelectedLines';
-export * from './useGetItemPricing';
```
Comment (lache-melvin) [inline client/packages/invoices/src/OutboundShipment/api/hooks/utils/index.ts:8]: We just include pricing info in the outbound line itemData query

File: client/packages/invoices/src/Prescriptions/api/hooks/utils.tsx [hunk 10]
```diff
@@ -157,25 +156,10 @@ export interface DraftPrescriptionLineSeeds {
 export const createDraftPrescriptionLineFromStockLine = ({
   invoiceId,
   stockLine,
-  defaultPricing,
```
Comment (lache-melvin) [inline client/packages/invoices/src/Prescriptions/api/hooks/utils.tsx:160]: default pricing was never being passed in for prescriptions...

File: client/packages/invoices/src/StockOut/Components/StockOutAlerts.tsx [hunk 11]
```diff
@@ -24,7 +24,15 @@ export const StockOutAlerts = ({
       ? allocationAlerts
       : [];
 
-  if (alerts.length === 0) return null;
+  return <AllocationAlerts allocationAlerts={alerts} />;
```
Comment (lache-melvin) [inline client/packages/invoices/src/StockOut/Components/StockOutAlerts.tsx:27]: leaving existing component for now, for prescriptions. using the display only one below for outbound

File: client/packages/invoices/src/StockOut/utils.tsx [hunk 12]
```diff
@@ -177,268 +121,6 @@ export const getAllocatedQuantity = (draftStockOutLines: DraftStockOutLine[]) =>
     3
   );
 
-export const issueStock = (
```
Comment (lache-melvin) [inline client/packages/invoices/src/StockOut/utils.tsx:180]: `allocation` folder should prob get moved here to be shared with prescription, but for now seeing as only used in outbound, kept it there, along with its utils. prescriptions currently has its own copy of these functions

File: server/graphql/invoice_line/src/mutations/outbound_shipment_line/save_item_lines.rs [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+fn map_error(error: SaveOutboundShipmentLinesError) -> Result<InvoiceNode> {
+    use SaveOutboundShipmentLinesError::*;
+    let formatted_error = format!("{:#?}", error);
+
+    log::error!("Error: {}", formatted_error);
+
+    // Future TODO: Implement structured errors where needed
+    // (Would only occur if 2 people editing at same time)
```
Comment (lache-melvin) [inline server/graphql/invoice_line/src/mutations/outbound_shipment_line/save_item_lines.rs:92]: Were returning all the structured errors in the old mutation - but weren't handling them on frontend? So I figured this was simpler, until there are actually errors we want to translate!

File: server/graphql/types/src/types/draft_outbound_shipment_line.rs [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    ) -> Vec<DraftOutboundShipmentLineNode> {
+        shipment_lines
+            .into_iter()
+            .map(|shipment_line| DraftOutboundShipmentLineNode { shipment_line })
+            .collect()
+    }
+}
+
+#[Object]
+impl DraftOutboundShipmentLineNode {
```
Comment (lache-melvin) [inline server/graphql/types/src/types/draft_outbound_shipment_line.rs:40]: custom node for the modal, just the fields we need

File: server/graphql/types/src/types/draft_outbound_shipment_line.rs [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+#[Object]
+impl DraftOutboundShipmentItemData {
+    pub async fn draft_lines(&self) -> Vec<DraftOutboundShipmentLineNode> {
+        DraftOutboundShipmentLineNode::from_vec(self.lines.clone())
+    }
+
+    pub async fn placeholder_quantity(&self) -> Option<f64> {
+        self.placeholder_quantity
+    }
```
Comment (lache-melvin) [inline server/graphql/types/src/types/draft_outbound_shipment_line.rs:21]: Returning placeholder quan separately - draft lines is therefore just the StockOut lines. easier on logic that constantly filtering for a placeholder...

**Comment by andreievg** [comment]
> @mark-prins I have one concern here... as discussed, removed the pack based auto-allocation, it's just in units now.
> 
> But if I'm looking at an item where pack size is 1000, and I want to ship 100 packs... I'd have to type 100,000?
> 
> I know the "any" thing was confusing, but I'm wondering if the "if all pack sizes the same, issue in that pack size" was somehting that made sense? Though, I guess its also confusing if sometimes the modal is in packs, and sometimes units.
> 
> Again, as the requirements here seem fuzzy, I don't want to spend too much time on what's not useful, but wanted to raise with you before we deploy 2.8, in case this is something we want to smooth over before then: 🙏

DJ would 100% need either default pack size or something like that, but maybe allocation logic is separate to "calculation aid" (for selecting if it's dose/unit/particular pack that is entere in 'issue'
)

File: client/packages/common/src/utils/arrays/SortUtils.ts (hunk 1 above)
Comment (andreievg) [reply client/packages/common/src/utils/arrays/SortUtils.ts:29]: > "non-expiring" batches should be allocated last

Yep that's how it's done on b/e

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/useAllocationContext.ts [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+    const sortedLines = draftLines.sort(SorterByStrategy[strategy]);
+
+    // Separate lines here, so only dealing with allocatable lines going forward
+    // Note - expired is still considered allocatable, just not via auto-allocation
+    const [allocatableLines, nonAllocatableLines] = ArrayUtils.partition(
+      sortedLines,
+      line =>
+        scannedBatch
+          ? scannedBatchFilter(sortedLines, line, scannedBatch)
+          : canAllocate(line)
```
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/useAllocationContext.ts:89]: canAllocate is also checked in scannedBatchFilter, wonder if it should be lifted up

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/useAllocationContext.ts [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+      isDirty: false,
+      draftLines: [],
+      nonAllocatableLines: [],
+      placeholderQuantity: null,
+      initialisedForItemId: null,
+      allocateIn: AllocateIn.Units,
+      alerts: [],
+    })),
+
+  setDraftLines: lines =>
```
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/useAllocationContext.ts:116]: I wasn't sure if this is only used internally, if it is could move it up to be a closure method (just cause on initialise lines are sorted, but could set unsorted lines with this one)

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/allocateQuantities.ts [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+ * with a bias towards the first stock line(s) in the list (e.g. first expiring).
+ *
+ * This can take up to 3 steps, let's look at an example with 2 stock lines:
+ *
+ * 200 x pack size 1    (exp 01/26)
+ * 3 x   pack size 100  (exp 03/26)
+ *
+ * And we are requesting 350 units.
+ *
+ * Step 1: from top of list allocate to the nearest (rounded down) pack size
```
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/allocateQuantities.ts:22]: We actually need to confirm this allocation, I think OG uses FEFO full stop, with fractional number of packs if below pack size.Which makes sense, imagine you have a pallet of good expiring this week, and a carton expiring this year, system should really suggest that you split the pallet when issuing a carton

File: client/packages/invoices/src/StockOut/utils.tsx [hunk 19]
```diff
@@ -177,268 +121,6 @@ export const getAllocatedQuantity = (draftStockOutLines: DraftStockOutLine[]) =>
...
-      htmlInput: {
-        sx: {
-          backgroundColor: props.isDisabled ? undefined : 'background.white',
-        },
-      },
-    }}
-  />
-);
-
 export const getAllocationAlerts = (
```
Comment (andreievg) [inline client/packages/invoices/src/StockOut/utils.tsx:124]: I like this as one place to figure see what alerts will trigger and when

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/useAllocationContext.ts [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+        placeholderQuantity:
+          placeholderQuantity === null ? null : stillToAllocate,
+      }));
+    }
+  },
+
+  manualAllocate: (lineId, quantity) => {
+    const { draftLines, setDraftLines } = get();
+
+    const updatedLines = issueStock(draftLines, lineId, quantity);
```
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/useAllocationContext.ts:181]: With this need allocateIn (or is that in future PR)

Comment (lache-melvin) [reply client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/useAllocationContext.ts:181]: Yah future PR

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/useAllocationContext.ts [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+        hasExpired,
+        format,
+        t
+      );
+
+      set(state => ({
+        ...state,
+        alerts,
+        placeholderQuantity:
+          placeholderQuantity === null ? null : stillToAllocate,
```
[the remark this answers is not in the document]
Comment (lache-melvin) [reply client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/useAllocationContext.ts:173]: Shouldn't set the placeholder quantity if not using a placeholder line? After allocated status, we should just under-allocate if we don't have enough stock..

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/utils.ts [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+  draftLines: DraftStockOutLineFragment[],
+  idToIssue: string,
+  packs: number
+) => {
+  const foundRowIdx = draftLines.findIndex(({ id }) => id === idToIssue);
+  const foundRow = draftLines[foundRowIdx];
+  if (!foundRow) return draftLines;
+
+  const newDraftLines = [...draftLines];
+  newDraftLines[foundRowIdx] = {
```
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/utils.ts:44]: I feel like this is a mutation ? (since previous restructure doesn't copy draftLines, it just create a new array with them), I think for immutablility have to destructure down to primatives for objects that change, so somthing like [...sliceBefore, { ...found, newNumberOfPacks} ...sliceAfter]

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/useOpenedWithBarcode.ts [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+};
+
+const useFocusNumberOfPacksInput = (
+  batch: string | undefined,
+  itemId: string | null
+) => {
+  useEffect(() => {
+    if (!batch || !itemId) return;
+    setTimeout(() => {
+      const input = document.getElementById(getPackQuantityCellId(batch));
```
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/useOpenedWithBarcode.ts:45]: I guess this is copy and paste ? useRef set to zustand state would be the alternative

**Review by jmbrunskill — COMMENTED** [review]
Nice! Great effort lots of code here.
I'm having setup issues testing it (pretty sure it's nothing to do with your branch) so adding a comment and will try to test and approve later.

**Review by jmbrunskill — APPROVED** [review]
Ok did some testing looks good.

Haven't tested on Tablet or with barcodes though. Has that been done?

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx [hunk 24]
```diff
@@ -1,197 +1,130 @@
...
+  const {
+    draftLines,
+    allocatedUnits,
+    placeholderQuantity,
+    alerts,
+    isDirty,
+    setAlerts,
+    clear,
+  } = useAllocationContext(state => ({
+    draftLines: state.draftLines,
```
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx:69]: not sure if this selector is really needed

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx [hunk 25]
```diff
@@ -1,197 +1,130 @@
...
-    if (showZeroQuantityConfirmation && newVal !== 0)
-      setShowZeroQuantityConfirmation(false);
-
-    return newAllocateQuantities;
-  };
-
-  const canAutoAllocate = !!(currentItem && draftStockOutLines.length);
-  const okNextDisabled =
-    (mode === ModalMode.Update && nextDisabled) || !currentItem;
+  const okNextDisabled = (mode === ModalMode.Update && nextDisabled) || !itemId;
```
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx:95]: I was confused about this (although I think it's product decision, this behaviour will need to be listed to confirm that's what we want), in add new can ok next to new item but in update mode cant

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx [hunk 26]
```diff
@@ -1,197 +1,130 @@
...
-        setCurrentItem(next);
+        setItemId(next.id);
         return true;
       }
       if (mode === ModalMode.Create) {
-        setCurrentItem(null);
+        setItemId(undefined);
         return true;
       }
       onClose();
```
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx:130]: don't think this ever happens

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/Allocation.tsx (hunk 4 above)
Comment (andreievg) [reply client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/Allocation.tsx:45]: This one confused me, Allocation will 'refresh' because loading will render while fetching. It makes sense once you look at it in more detail, but I still think we can improve on that at some point, almost feel like when item is fetched there is a bunch of things to be done like this, and kind of feels like it should be an await.
This is still readable, i'll just put a comment about this useEffect

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/utils.ts [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+  draftLines: DraftStockOutLineFragment[];
+  placeholderQuantity: number | null;
+}) =>
+  NumUtils.round(
+    draftLines.reduce(
+      (acc, { numberOfPacks, packSize }) => acc + numberOfPacks * packSize,
+      0
+    ),
+    3
+  ) + (placeholderQuantity ?? 0);
```
Comment (andreievg) [inline client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/utils.ts:32]: This is another one to add to description of behaviour, if we have placeholder less then what you can supply, and we manually supply that, placeholder stays. We kind of need something similar to prescription, 'prescribed' quantity or it could be placeholder, but the rule is, placeholder is populated with how much you want to issue, but is displayed as placholder quantity - how much you've issued

File: server/service/src/invoice_line/get_draft_outbound_lines.rs [hunk 28]
```diff
@@ -0,0 +1,336 @@
...
+
+    let available_stock_lines = stock_line_repo.query_by_filter(
+        StockLineFilter::new()
+            // For selected item
+            .item_id(EqualFilter::equal_to(item_id))
+            // With available stock in this store
+            .store_id(EqualFilter::equal_to(&store_id))
+            .has_packs_in_store(true)
+            // Exclude stock lines already in the invoice
+            .id(EqualFilter::not_equal_all(existing_stock_line_ids)),
```
[the remark this answers is not in the document]
Comment (andreievg) [reply server/service/src/invoice_line/get_draft_outbound_lines.rs:122]: James noticed this while i was passing by, on the first look yeah it seems like it should be available, but since we are generating draft lines, it's actually total (we show total in item selection too), the idea is you have some in stock even though it might not be available, and the user can then figure out if they want to re-locate from other shipment

File: server/service/src/invoice_line/get_draft_outbound_lines.rs [hunk 29]
```diff
@@ -0,0 +1,336 @@
...
+            number_of_packs,
+            stock_line_id,
+            pack_size,
+            batch,
+            expiry_date,
+            location_id,
+            sell_price_per_pack,
+            // Stock already included in the invoice wouldn't be on the stock line available packs,
+            // but should be considered available in the context of this invoice
+            available_packs: available_number_of_packs + number_of_packs,
```
Comment (andreievg) [inline server/service/src/invoice_line/get_draft_outbound_lines.rs:213]: nice

File: server/service/src/invoice_line/save_outbound_shipment_item_lines/generate.rs [hunk 30]
```diff
@@ -0,0 +1,153 @@
...
+    let check_already_exists = |id: &str| {
+        existing_lines
+            .iter()
+            .any(|line| line.invoice_line_row.id == id)
+    };
+
+    let lines_to_add = lines
+        .clone()
+        .into_iter()
+        .filter(|line| line.number_of_packs > 0.0 && !check_already_exists(&line.id))
```
Comment (andreievg) [inline server/service/src/invoice_line/save_outbound_shipment_item_lines/generate.rs:54]: hmm, so for the 'ids'for draft lines, I think we should really go with stock_line_ids (when getting draft lines and here when checking if there is invoice line)

File: server/service/src/invoice_line/save_outbound_shipment_item_lines/generate.rs [hunk 31]
```diff
@@ -0,0 +1,153 @@
...
+                batch: None,
+                pack_size: None,
+                expiry_date: None,
+                prescribed_quantity: None,
+                total_before_tax: None,
+                tax_percentage: None,
+                note: None,
+                location_id: None,
+                cost_price_per_pack: None,
+                sell_price_per_pack: None,
```
Comment (andreievg) [inline server/service/src/invoice_line/save_outbound_shipment_item_lines/generate.rs:76]: This doesn't feel right, I think sell price needs to be set ?

File: server/service/src/invoice_line/save_outbound_shipment_item_lines/generate.rs [hunk 32]
```diff
@@ -0,0 +1,153 @@
...
+                quantity: quantity as u32, // worth cleaning up this input... gets saved as f64?
+            })
+        }
+        _ => ManagePlaceholderLine::NothingToDo,
+    };
+
+    Ok(GenerateResult {
+        lines_to_add,
+        lines_to_update,
+        lines_to_delete,
```
Comment (andreievg) [inline server/service/src/invoice_line/save_outbound_shipment_item_lines/generate.rs:150]: I think we would still need to check (I think it's ok in carry over or cool down):
* Duplicated stock lines
* Stock lines in store
* Stock line belongs to that item
* Trip all invoice_lines that are not in the payload (for that item)

**Comment by andreievg** [comment]
What a massive effort !

Thanks @lache-melvin, i went through it all and did some light testing, i think the only big question was about pricing changes in prescriptions. 

It's a huge improvement, thank you, I feel more confident navigating this part of the code base now : )

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/useOpenedWithBarcode.ts (hunk 23 above)
Comment (lache-melvin) [reply client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/hooks/useOpenedWithBarcode.ts:45]: Yeah this was existing...

File: server/graphql/lib.rs [hunk 33]
```diff
@@ -1,3 +1,5 @@
+#![recursion_limit = "256"]
```
[the remark this answers is not in the document]
Comment (lache-melvin) [reply server/graphql/lib.rs:1]: Ah yeah sorry, compilation failed... I've seen this added in a couple places now (it's the recommended fix by the compiler?)

Might raise an issue to review it though, along with other developer experience performance things...

File: server/service/src/invoice_line/get_draft_outbound_lines.rs [hunk 34]
```diff
@@ -0,0 +1,336 @@
...
+    invoice_id: &str,
+) -> Result<
+    (
+        Vec<DraftOutboundShipmentLine>,
+        Option<f64>, /* placeholder_quantity */
+    ),
+    ListError,
+> {
+    let outbound = get_invoice(ctx, Some(&store_id), invoice_id)?.ok_or(
+        ListError::DatabaseError(RepositoryError::DBError {
```
[the remark this answers is not in the document]
Comment (lache-melvin) [reply server/service/src/invoice_line/get_draft_outbound_lines.rs:44]: Yeah not found with a message would be good, especially when its otherwise not clear which record wasn't found 😁

File: client/packages/invoices/src/Prescriptions/api/hooks/utils.tsx [hunk 35]
```diff
@@ -184,7 +168,7 @@ export const createDraftPrescriptionLineFromStockLine = ({
     prescribedQuantity: 0,
     location: stockLine?.location,
     expiryDate: stockLine?.expiryDate,
-    sellPricePerPack,
+    sellPricePerPack: stockLine?.sellPricePerPack ?? 0,
```
[the remark this answers is not in the document]
Comment (lache-melvin) [reply client/packages/invoices/src/Prescriptions/api/hooks/utils.tsx:171]: Yep sellPrice was using logic as though a pricing object was being passed in, but we were never calling this function with a pricing object. So just stripped back, so behaviour was clear

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx (hunk 25 above)
Comment (lache-melvin) [reply client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/OutboundLineEdit.tsx:95]: Added to carry over, behaviour to list and confirm

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/utils.ts (hunk 22 above)
Comment (lache-melvin) [reply client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/utils.ts:44]: Ooh yeah great point, won't be entirely new objects, this was from existing logic... and it does look like everything is updating ok, maybe as we aren't doing any deeper comparisons anywhere?

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/utils.ts [hunk 36]
```diff
@@ -0,0 +1,73 @@
...
+export const scannedBatchFilter = (
+  allLines: DraftStockOutLineFragment[],
+  selectedLine: DraftStockOutLineFragment,
+  scannedBatch: string
+) => {
+  if (!canAllocate(selectedLine)) return false;
+
+  const linesIncludeScannedBatch = allLines.some(l => l.batch === scannedBatch);
+
+  // If the requested batch is not in the list, we can allocate any line
```
[the remark this answers is not in the document]
Comment (lache-melvin) [reply client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/allocation/utils.ts:69]: Yes this is same as existing... will document for confirmation though, its not super intuitive to me...

File: server/service/src/invoice_line/save_outbound_shipment_item_lines/generate.rs (hunk 31 above)
Comment (lache-melvin) [reply server/service/src/invoice_line/save_outbound_shipment_item_lines/generate.rs:76]: Sell price isn't editable in outbound, so leaving as none here means the stock_out line will just use whatever the sell price is on the stock line

**Comment by lache-melvin** [comment]
Thanks for review both of you!! I will merge this (FYI I think will do without squashing seeing as we have a few dependent PRs 😅) - am creating a couple of follow up issues to capture the other comments in here 🙏

File: client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/Allocation.tsx (hunk 4 above)
Comment (lache-melvin) [reply client/packages/invoices/src/OutboundShipment/DetailView/OutboundLineEdit/Allocation.tsx]: d3dd097cec
